packages/cli/src/tools/releaseChecker/getLatestRelease.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @flow
+ */
+import https from 'https';
+import semver from 'semver';
+import logger from '../logger';
+import cacheManager from './releaseCacheManager';
+
+export type Release = {
+  version: string,
+  changelogUrl: string,
+};
+
+/**
+ * Checks via GitHub API if there is a newer stable React Native release and,
+ * if it exists, returns the release data.
+ *
+ * If the latest release is not newer or if it's a prerelease, the function
+ * will return undefined.
+ */
+export default (async function(currentVersion: string) {
+  logger.debug('Checking for a newer version of React Native');
+  try {
+    logger.debug(`Current version: ${currentVersion}`);
+
+    const cachedLatest: string = cacheManager.get('latestVersion');
+
+    if (cachedLatest) {
+      logger.debug(`Cached release version: ${cachedLatest}`);
+    }
+
+    logger.debug('Checking for newer releases on GitHub');
+    const eTag = cacheManager.get('eTag');
+    const latestVersion = await getLatestRnDiffPurgeVersion(eTag);
+    logger.debug(`Latest release: ${latestVersion}`);
+
+    if (
+      semver.compare(latestVersion, currentVersion) === 1 &&
+      !semver.prerelease(latestVersion)
+    ) {
+      return {
+        version: latestVersion,
+        changelogUrl: buildChangelogUrl(latestVersion),
+      };
+    }
+  } catch (e) {
+    logger.debug(
+      'Something went wrong with remote version checking, moving on',
+    );
+  }
+});
+
+function buildChangelogUrl(version: string) {
+  return `https://github.com/facebook/react-native/releases/tag/v${version}`;
+}
+
+/**
+ * Returns the most recent React Native version available to upgrade to.
+ */
+async function getLatestRnDiffPurgeVersion(eTag: ?string): Promise<string> {
+  const options = {
+    hostname: 'api.github.com',
+    path: '/repos/react-native-community/rn-diff-purge/tags',
+    // https://developer.github.com/v3/#user-agent-required
+    headers: ({'User-Agent': 'React-Native-CLI'}: Headers),
+  };
+
+  if (eTag) {
+    options.headers['If-None-Match'] = eTag;
+  }
+
+  const response = await httpsGet(options);
+
+  // Remote is newer.
+  if (response.statusCode === 200) {
+    const latestVersion = JSON.parse(response.body)[0].name.substring(8);
+
+    // Update cache only if newer release is stable.
+    if (!semver.prerelease(latestVersion)) {
+      logger.debug(`Saving ${response.eTag} to cache`);
+      cacheManager.set('eTag', response.eTag);
+      cacheManager.set('latestVersion', latestVersion);
+    }
+
+    return latestVersion;
+  }
+
+  // Cache is still valid.
+  if (response.statusCode === 304) {
+    return cacheManager.get('latestVersion');
+  }
+
+  // Should be returned only if something went wrong.
+  return '0.0.0';
+}
+
+type Headers = {
+  'User-Agent': string,
+  [header: string]: string,
+};
+
+function httpsGet(options: any) {
```

Comment:
We already have a helper for that: https://github.com/react-native-community/cli/blob/master/packages/cli/src/tools/fetch.js please use it :)

Reply:
That implementation is a bit too basic for my use case: it rejects for status codes `300` and above with only an error and nothing else. However, in my case we are also interested in `304`s and result headers, so it wouldn't really work without first augmenting that `fetch` implementation.

Reply:
But, I'd be glad if you had ideas on how to unify both functions in a followup :)

Reply:
I'd be glad to. I should also point out that we are also using `node-fetch` [here](https://github.com/react-native-community/cli/blob/master/packages/tools/src/isPackagerRunning.js), adding up to three different used solutions 🙈.

Reply:
Oh, good catch! We should kill it. I didn't notice that we don't have `node-fetch` in our dependencies so it works because of luck. cc @lucasbento